DDG4/edm4hep/EDM4hepFileReader.cpp

```diff
@@ -0,0 +1,320 @@
+ *
+ @{
+ \package EDM4hepFileReader
+ * \brief Plugin to read EDM4hep files
+ *
+ *
+ @}
+*/
+
+#include <CLHEP/Units/SystemOfUnits.h>
+
+#include <DDG4/EventParameters.h>
+#include <DDG4/Factories.h>
+#include <DDG4/Geant4InputAction.h>
+#include <DDG4/RunParameters.h>
+
+#include <edm4hep/EventHeaderCollection.h>
+#include <edm4hep/MCParticleCollection.h>
+
+#include <podio/Frame.h>
+#include <podio/ROOTReader.h>
+
+typedef dd4hep::detail::ReferenceBitMask<int> PropertyMask;
+
+
+namespace dd4hep::sim {
+
+  // we use the index of the objectID to identify particles
+  // we will not support MCParticles from different collections
+  using MCPARTICLE_MAP=std::map<int, int>;
+
+  /// get the parameters from the GenericParameters of the input EDM4hep frame and store them in the EventParameters
+  /// extension
+  template <class T=podio::GenericParameters> void EventParameters::ingestParameters(T const& source) {
+    auto const& intKeys = source.template getKeys<int>();
+    for(auto const& key: intKeys) {
+      m_intValues[key] = source.template get<std::vector<int>>(key).value();
+    }
+    auto const& floatKeys = source.template getKeys<float>();
+    for(auto const& key: floatKeys) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    auto const& doubleKeys = source.template getKeys<double>();
+    for(auto const& key: doubleKeys) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    using std::string;
+    auto const& stringKeys = source.template getKeys<string>();
+    for(auto const& key: stringKeys) {
+      m_strValues[key] = source.template get<std::vector<string>>(key).value();
+    }
+  }
+
+  /// get the parameters from the GenericParameters of the input EDM4hep run frame and store them in the RunParameters
+  /// extension
+  template <class T=podio::GenericParameters> void RunParameters::ingestParameters(T const& source) {
+    auto const& intKeys = source.template getKeys<int>();
+    for(auto const& key: intKeys) {
+      m_intValues[key] = source.template get<std::vector<int>>(key).value();
+    }
+    auto const& floatKeys = source.template getKeys<float>();
+    for(auto const& key: floatKeys) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    auto const& doubleKeys = source.template getKeys<double>();
+    for(auto const& key: doubleKeys) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    using std::string;
+    auto const& stringKeys = source.template getKeys<string>();
+    for(auto const& key: stringKeys) {
+      m_strValues[key] = source.template get<std::vector<string>>(key).value();
+    }
+  }
+
+  /// Class to read EDM4hep files
+  /**
+   *  \version 1.0
+   *  \ingroup DD4HEP_SIMULATION
+   */
+  class EDM4hepFileReader : public Geant4EventReader {
+  protected:
+    /// Reference to reader object
+    podio::ROOTReader m_reader {};
```

Comment:
```suggestion
    podio::Reader m_reader {};
```
Would immediately enable support for more backends (most importantly RNTuple, to a lesser extent SIO), without having to do that later.

Reply:
What is the file extension convention for sio files?

Reply:
`"sio"` (as per [this check](https://github.com/AIDASoft/podio/blob/d3f03813d7bf50f87c3aa4f32e619cebeb060113/src/Reader.cc#L72)

Reply:
Added sio extension